### PR TITLE
Refactor extension_providers_config

### PR DESCRIPTION
**Pull request by IbraAoad**
## Issue
To enable workload tracing in Istio, we currently append an `opentelemetry` configuration to the list of extension providers generated by `istioctl`. However, this list is currently hard-coded to only accept the tracing provider.

## Solution
This PR refactors the relevant code to allow adding extra providers easily without hard-coding their indices. It also introduces a mechanism to flatten the providers' configuration to match the `istioctl` format and adds the necessary scenario tests.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -261,38 +261,71 @@ def _get_gateway_apis_kubernetes_resource_manager(self):
...
-            f"//{workload_tracing_endpoint}"
-        ).hostname
-        tracing_config["meshConfig.defaultProviders.tracing[0]"] = "otel-tracing"
-        tracing_config["meshConfig.defaultConfig.tracing.sampling"] = 100.0
-
-        return tracing_config
+    def _workload_tracing_provider(self) -> Tuple[List[Any], Dict[str, Any]]:
+        """Return a tuple with a list containing the tracing provider and global tracing settings."""
+        providers = []
+        global_config = {}
```
Comment (ca-scribner) [inline src/charm.py:267]: nit: works fine like this, but since you only have one provider and one config to return, making the objs here and .appending them later is just extra code

File: src/charm.py [hunk 2]
```diff
@@ -261,38 +261,71 @@ def _get_gateway_apis_kubernetes_resource_manager(self):
...
+        if self.workload_tracing.is_ready():
+            workload_tracing_endpoint = self.workload_tracing.get_endpoint("otlp_grpc")
```
Comment (ca-scribner) [inline src/charm.py:269]: nit: get_endpoint does an is_ready and returns None if not ready
```suggestion
        if workload_tracing_endpoint := self.workload_tracing.get_endpoint("otlp_grpc"):  # will be false if tracing is not ready because get_endpoint returns None
```

File: src/charm.py [hunk 3]
```diff
@@ -261,38 +261,71 @@ def _get_gateway_apis_kubernetes_resource_manager(self):
...
+            for k, v in value.items():
+                new_prefix = f"{prefix}.{k}" if prefix else k
+                flat.update(self._flatten_config(v, new_prefix))
+        elif isinstance(value, list):
+            for i, item in enumerate(value):
+                new_prefix = f"{prefix}[{i}]"
+                flat.update(self._flatten_config(item, new_prefix))
+        else:
+            flat[prefix] = value
+        return flat
```
Comment (ca-scribner) [inline src/charm.py:301]: We dont use `self` here - extract this to be a helper.  It'll be way easier to unit test that way too

Comment (ca-scribner) [inline src/charm.py:301]: I think there's a bug here, but I can't tell.  Some unit testing will prove it right or wrong.

Why I think there's a bug is that it does `flat[prefix] = value`.  That will fail if `prefix=""`.  For this particular use case i dont think you'll have `prefix=""`, but in general you could (`_flatten_config({"a": {"b": "c"}}, prefix="")` should be valid

This'll be really easy to figure out with a few unit tests though.  Make a basic test with a good set of parameters to catch the edge cases and you'll know for sure

File: src/charm.py [hunk 4]
```diff
@@ -261,38 +261,71 @@ def _get_gateway_apis_kubernetes_resource_manager(self):
...
+        return providers, global_config
+
+    def _flatten_config(self, value: Any, prefix: str = "") -> Dict[str, Any]:
+        """Recursively flatten a nested dictionary or list into a dictionary of key/value pairs."""
+        flat: Dict[str, Any] = {}
+        if isinstance(value, dict):
+            for k, v in value.items():
+                new_prefix = f"{prefix}.{k}" if prefix else k
+                flat.update(self._flatten_config(v, new_prefix))
+        elif isinstance(value, list):
```
Comment (ca-scribner) [inline src/charm.py:295]: Probably we can include tuples here too
```suggestion
        elif isinstance(value, (list, tuple)):  # I think this is the syntax?
```

File: src/charm.py (hunk 3 above)
Comment (ca-scribner) [reply src/charm.py:301]: Write it however you like, but I think it'll be easier to write if the return is a list of (key, value) tuples rather than a dict.  But if what you have works and you like it better, lgtm :) 

The reason I think about the list of tuples is I think it removes the need to pass prefixes all the way down.  The calling function can just recurse with something like `flat.extend([(prefix + k, v) for k, v in flatten_config(value)`. 
 But if that makes things harder, ignore me :D  Just an idea in case it helps

**Review by ca-scribner — COMMENTED** [review]
Directionally I approve of this.  I think there's a bug or two, but I'm not sure (unit testing will confirm it works).  But I love the approach

File: src/charm.py (hunk 2 above)
Comment (IbraAoad) [reply src/charm.py]: Apparently we can't do this without a guard 

```
  File "/var/lib/juju/agents/unit-istio-k8s-0/charm/./src/charm.py", line 291, in _workload_tracing_provider
    if not (endpoint := self.workload_tracing.get_endpoint("otlp_grpc")):
  File "/var/lib/juju/agents/unit-istio-k8s-0/charm/lib/charms/tempo_coordinator_k8s/v0/tracing.py", line 931, in get_endpoint
    raise ProtocolNotRequestedError(protocol, relation)
charms.tempo_coordinator_k8s.v0.tracing.ProtocolNotRequestedError: ('otlp_grpc', None)
```

**Review by dstathis — APPROVED** [review]

File: src/charm.py (hunk 2 above)
Comment (ca-scribner) [reply src/charm.py]: hmm odd.  ok!

File: src/charm.py (hunk 3 above)
Comment (ca-scribner) [reply src/charm.py]: ty for tests, lgtm!  also, I never really thought about how `""` is a valid key for a dict...

File: tests/scenario/test_extension_providers.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+#############################################
+# Tests for _build_extension_providers_config
+#############################################
+
+
+def test_build_extension_providers_config_empty(istio_core_context):
+    """Test that an empty providers list returns an empty configuration."""
+    state = State(relations=[])
+    with istio_core_context(istio_core_context.on.config_changed(), state) as mgr:
```
Comment (ca-scribner) [inline tests/scenario/test_extension_providers.py:138]: nit: since these tests dont use anything from the event handling, I'd use update-status (since we do less work on those events and its less likely to interfere).  But this works too

File: tests/scenario/test_extension_providers.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    """Verify that nested dictionaries are recursively flattened with the provided prefix."""
+    data = {"a": {"b": 1, "c": {"d": "x"}}, "e": 2}
+    prefix = "prefix"
+    expected = {
+        "prefix.a.b": 1,
+        "prefix.a.c.d": "x",
+        "prefix.e": 2,
+    }
+    result = flatten_config(data, prefix)
+    assert result == expected
```
Comment (ca-scribner) [inline tests/scenario/test_extension_providers.py:36]: Any reason why we don't use `pytest.mark.parametrize` here?  This looks like 3 parametrized tests rather than 13 independent tests

Using the top two as an example:
```suggestion
@pytest.mark.parametrize(
  "data, prefix, expected",
  [
    # flat dict
    (
      {"a": "value1", "b": "value2"}, 
      "provider[0]",
      {
        "provider[0].a": "value1",
        "provider[0].b": "value2",
      },
    ),
    # nested dict
    (
      {"a": {"b": 1, "c": {"d": "x"}}, "e": 2}, 
      "prefix",
      {
        "prefix.a.b": 1,
        "prefix.a.c.d": "x",
        "prefix.e": 2,
      },
    ),
    # ...
  ]
)    
def test_flatten_config(data, prefix, expected):
    result = flatten_config(data, prefix)
    assert result == expected

```

the scenario tests below are the same, too

File: tests/scenario/test_extension_providers.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+from scenario import State
+
+from charm import IstioCoreCharm, flatten_config
+
+#############################################
+# Tests for flatten_config
+#############################################
+
+
+def test_flatten_config_with_flat_dict():
```
Comment (ca-scribner) [inline tests/scenario/test_extension_providers.py:14]: nit: I'd put this top group in tests/unit/test_extension_providers.py since they aren't scenario tests

**Review by ca-scribner — APPROVED** [review]
lgtm, with some suggestions to reduce the code a bit

File: tests/scenario/test_extension_providers.py (hunk 6 above)
Comment (IbraAoad) [reply tests/scenario/test_extension_providers.py]: I personally try to refrain from using `pytest.mark.parametrize` just because I like how separate small tests are more readable and have clearer descriptions and code reduction isn't really adding anything to overall performance, I changed them all tho to use `parametrize` to stay consistent with our charms  :angel:
